dbt/adapters/databricks/credentials.py

```diff
@@ -238,172 +245,174 @@ def extract_cluster_id(cls, http_path: str) -> Optional[str]:
+    token: Optional[str] = None
+    auth_type: Optional[str] = None
 
-        oauth_client = OAuthClient(
-            host=self.host or "",
-            client_id=self.client_id or CLIENT_ID,
-            client_secret="",
-            redirect_url=self.oauth_redirect_url or REDIRECT_URL,
-            scopes=self.oauth_scopes or SCOPES,
+    @classmethod
+    def create_from(cls, credentials: DatabricksCredentials) -> "DatabricksCredentialManager":
+        return DatabricksCredentialManager(
+            host=credentials.host or "",
+            token=credentials.token,
+            client_id=credentials.client_id or CLIENT_ID,
+            client_secret=credentials.client_secret or "",
+            azure_client_id=credentials.azure_client_id,
+            azure_client_secret=credentials.azure_client_secret,
+            oauth_redirect_url=credentials.oauth_redirect_url or REDIRECT_URL,
+            oauth_scopes=credentials.oauth_scopes or SCOPES,
+            auth_type=credentials.auth_type,
+        )
+
+    def authenticate_with_pat(self) -> Config:
+        return Config(
+            host=self.host,
+            token=self.token,
+        )
+
+    def authenticate_with_oauth_m2m(self) -> Config:
+        return Config(
+            host=self.host,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            auth_type="oauth-m2m",
+        )
+
+    def authenticate_with_external_browser(self) -> Config:
+        return Config(
+            host=self.host,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            auth_type="external-browser",
         )
 
-        return SessionCredentials.from_dict(
-            client=oauth_client, raw=self._credentials_provider or {"token": {}}
+    def legacy_authenticate_with_azure_client_secret(self) -> Config:
+        return Config(
+            host=self.host,
+            azure_client_id=self.client_id,
+            azure_client_secret=self.client_secret,
+            auth_type="azure-client-secret",
         )
+
+    def authenticate_with_azure_client_secret(self) -> Config:
+        return Config(
+            host=self.host,
+            azure_client_id=self.azure_client_id,
+            azure_client_secret=self.azure_client_secret,
+            auth_type="azure-client-secret",
+        )
+
+    def __post_init__(self) -> None:
+        self._lock = threading.Lock()
+        with self._lock:
+            if not hasattr(self, "_config"):
+                self._config: Optional[Config] = None
+            if self._config is not None:
+                return
+
+            if self.token:
+                self._config = self.authenticate_with_pat()
+            elif self.azure_client_id and self.azure_client_secret:
+                self._config = self.authenticate_with_azure_client_secret()
+            elif not self.client_secret:
+                self._config = self.authenticate_with_external_browser()
```

Comment:
I love this, it will make it so much easier to add new authentication types!

I wonder if that could be made even simpler by just having a common generic method to build the Databricks SDK for all authentication types except the legacy ones.

The logic would then be reversed: we first check whether the authentication method used is legacy and otherwise just run something like: `self._config = self.databricks_config`
```python
    @property
    def databricks_config(self) -> Config:
        return Config(
            host=self.host,
            client_id=self.client_id,
            client_secret=self.client_secret,
            auth_type=self.auth_type,
            azure_client_id=self.azure_client_id,
            azure_client_secret=self.azure_client_secret,
            token=self.token,
        )
```

I believe the `databricks.sdk.config.Config` class would handle the `None` values just fine by ignoring them. A major benefit of this approach, is that it would then be straightforward to pass through all of the attributes and thus enabling any possible authentication options.
It would also be easier to document and less confusing as you could just refer to Databricks' Unified Authentication / Python SDK authentication documentation. In addition it could also provide more flexibility by exposing additional configuration options.

For reference here is the full list of these attributes as of now:

- `host`
- `account_id`
- `token`
- `username`
- `password`
- `client_id`
- `client_secret`
- `profile`
- `config_file`
- `google_service_account`
- `google_credentials`
- `azure_workspace_resource_id`
- `azure_use_msi`
- `azure_client_secret`
- `azure_client_id`
- `azure_tenant_id`
- `azure_environment`
- `databricks_cli_path`
- `auth_type`
- `cluster_id`
- `warehouse_id`
- `serverless_compute_id`
- `skip_verify`
- `http_timeout_seconds`
- `debug_truncate_bytes`
- `debug_headers`
- `rate_limit`
- `retry_timeout_seconds`
- `metadata_service_url`
- `max_connection_pools`
- `max_connections_per_pool`
- `databricks_environment`